tests/http_spec.rs

```diff
@@ -0,0 +1,248 @@
+    Arc::new(server_context)
+  }
+}
+
+#[derive(Clone)]
+struct MockHttpClient {
+  upstream_mocks: Vec<(UpstreamRequest, UpstreamResponse)>,
+  expected_upstream_requests: Vec<UpstreamRequest>,
+}
+#[async_trait::async_trait]
+impl HttpClient for MockHttpClient {
+  async fn execute(&self, req: reqwest::Request) -> Result<Response, anyhow::Error> {
+    // Clone the mocks to allow iteration without borrowing issues.
+    let mocks = self.upstream_mocks.clone();
+
+    // Try to find a matching mock for the incoming request.
+    let mock = mocks
+      .iter()
+      .find(|(mock_req, _)| {
+        let method_match = req.method().as_str()
+          == serde_json::to_string(&mock_req.0.method.clone())
+            .expect("provided method is not valid")
+            .as_str()
+            .trim_matches('"');
+        let url_match = req.url().as_str() == mock_req.0.url.clone().as_str();
+        method_match && url_match
+      })
+      .expect("Mock not found");
+    // Assert upstream request
+    let upstream_request = mock.0.clone();
+    assert!(
+      self.expected_upstream_requests.contains(&upstream_request),
+      "Unexpected upstream request: {:?}",
+      upstream_request
+    );
+
+    // Clone the response from the mock to avoid borrowing issues.
+    let mock_response = mock.1.clone();
+
+    // Build the response with the status code from the mock.
+    let status_code = reqwest::StatusCode::from_u16(mock_response.0.status)?;
+    let mut response = Response { status: status_code, ..Default::default() };
+
+    // Insert headers from the mock into the response.
+    for (key, value) in mock_response.0.headers {
+      let header_name = HeaderName::from_str(&key)?;
+      let header_value = HeaderValue::from_str(&value)?;
+      response.headers.insert(header_name, header_value);
+    }
+
+    // Set the body of the response.
+    response.body = ConstValue::try_from(serde_json::from_value::<Value>(mock_response.0.body)?)?;
+
+    Ok(response)
+  }
+}
+
+async fn assert_downstream(spec: HttpSpec) {
+  for downstream_assertion in spec.downstream_assertions.iter() {
+    if let Some(Annotation::Fail) = spec.annotation {
+      let response = run(spec.clone(), &downstream_assertion).await.unwrap();
+      let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
+      assert_ne!(
+        body,
+        serde_json::to_string(&downstream_assertion.response.0.body).unwrap()
+      )
+    } else {
+      let response = run(spec.clone(), &downstream_assertion).await.unwrap();
+      let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
+      assert_eq!(
+        body,
+        serde_json::to_string(&downstream_assertion.response.0.body).unwrap()
+      )
+    }
+  }
+  log::info!("{} {} ... ok", spec.name, spec.path.display());
+}
+#[tokio::test]
+async fn test_body() -> std::io::Result<()> {
+  let spec = HttpSpec::cargo_read("tests/http").unwrap();
+  let spec = HttpSpec::filter_specs(spec);
```

Comment:
so, we actually building a test runner here 😄
I wonder if something like this https://crates.io/crates/datatest could help us with that

Reply:
They all use macro APIs, and I think we will end up using a lot of  statics also to share things between tests. I agree the current way isn't the best but I think we need a nicer library that can do more for us.